### PR TITLE
feat!: promote gke_backup_agent_config to ga

### PR DESCRIPTION
**Pull request by Tolsee**
* promote gke_backup_agent_config to ga

Fixes: https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/issues/1450

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
@Tolsee
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_generate_modules
 The modules need to be regenerated. Please run `make_build`. 
 ```bash 
Checking submodule's files generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/beta-private-cluster/README.md /tmp/tmp.SZ0ZN04jqW/workspace/modules/beta-private-cluster/README.md
183c183
< | gke\_backup\_agent\_config | Whether Backup for GKE agent is enabled for this cluster. | `bool` | `false` | no |
---
> | gke\_backup\_agent\_config | (Beta) Whether Backup for GKE agent is enabled for this cluster. | `bool` | `false` | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/beta-private-cluster/variables.tf /tmp/tmp.SZ0ZN04jqW/workspace/modules/beta-private-cluster/variables.tf
670c670
<   description = "Whether Backup for GKE agent is enabled for this cluster."
---
>   description = "(Beta) Whether Backup for GKE agent is enabled for this cluster."
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/beta-private-cluster/versions.tf /tmp/tmp.SZ0ZN04jqW/workspace/modules/beta-private-cluster/versions.tf
24c24
<       version = ">= 4.47.0, < 5.0"
---
>       version = ">= 4.45.0, < 5.0"
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/private-cluster/README.md /tmp/tmp.SZ0ZN04jqW/workspace/modules/private-cluster/README.md
168d167
< | gke\_backup\_agent\_config | Whether Backup for GKE agent is enabled for this cluster. | `bool` | `false` | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/private-cluster/cluster.tf /tmp/tmp.SZ0ZN04jqW/workspace/modules/private-cluster/cluster.tf
153,160d152
< 
<     dynamic "gke_backup_agent_config" {
<       for_each = local.gke_backup_agent_config
< 
<       content {
<         enabled = gke_backup_agent_config.value.enabled
<       }
<     }
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/private-cluster/main.tf /tmp/tmp.SZ0ZN04jqW/workspace/modules/private-cluster/main.tf
87,88d86
<   gke_backup_agent_config  = var.gke_backup_agent_config ? [{ enabled = true }] : [{ enabled = false }]
< 
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/private-cluster/variables.tf /tmp/tmp.SZ0ZN04jqW/workspace/modules/private-cluster/variables.tf
591,596d590
< 
< variable "gke_backup_agent_config" {
<   type        = bool
<   description = "Whether Backup for GKE agent is enabled for this cluster."
<   default     = false
< }
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/private-cluster/versions.tf /tmp/tmp.SZ0ZN04jqW/workspace/modules/private-cluster/versions.tf
24c24
<       version = ">= 4.47.0, < 5.0"
---
>       version = ">= 4.45.0, < 5.0"
Error: submodule's files generation has not been run, please run the
'make build' command and commit changes
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
modules/private-cluster/main.tf
--- old/modules/private-cluster/main.tf
+++ new/modules/private-cluster/main.tf
@@ -84,7 +84,7 @@
   }]
   cluster_gce_pd_csi_config = var.gce_pd_csi_driver ? [{ enabled = true }] : [{ enabled = false }]
 
-  gke_backup_agent_config  = var.gke_backup_agent_config ? [{ enabled = true }] : [{ enabled = false }]
+  gke_backup_agent_config = var.gke_backup_agent_config ? [{ enabled = true }] : [{ enabled = false }]
 
   cluster_authenticator_security_group = var.authenticator_security_group == null ? [] : [{
     security_group = var.authenticator_security_group
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
 ```

**Comment by comment-bot-dev** [comment]
@Tolsee
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_generate_modules
 The modules need to be regenerated. Please run `make_build`. 
 ```bash 
Checking submodule's files generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/beta-private-cluster/README.md /tmp/tmp.3unHo69Lk5/workspace/modules/beta-private-cluster/README.md
183c183
< | gke\_backup\_agent\_config | Whether Backup for GKE agent is enabled for this cluster. | `bool` | `false` | no |
---
> | gke\_backup\_agent\_config | (Beta) Whether Backup for GKE agent is enabled for this cluster. | `bool` | `false` | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/beta-private-cluster/variables.tf /tmp/tmp.3unHo69Lk5/workspace/modules/beta-private-cluster/variables.tf
670c670
<   description = "Whether Backup for GKE agent is enabled for this cluster."
---
>   description = "(Beta) Whether Backup for GKE agent is enabled for this cluster."
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/beta-private-cluster/versions.tf /tmp/tmp.3unHo69Lk5/workspace/modules/beta-private-cluster/versions.tf
24c24
<       version = ">= 4.47.0, < 5.0"
---
>       version = ">= 4.45.0, < 5.0"
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/private-cluster/README.md /tmp/tmp.3unHo69Lk5/workspace/modules/private-cluster/README.md
168d167
< | gke\_backup\_agent\_config | Whether Backup for GKE agent is enabled for this cluster. | `bool` | `false` | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/private-cluster/cluster.tf /tmp/tmp.3unHo69Lk5/workspace/modules/private-cluster/cluster.tf
153,160d152
< 
<     dynamic "gke_backup_agent_config" {
<       for_each = local.gke_backup_agent_config
< 
<       content {
<         enabled = gke_backup_agent_config.value.enabled
<       }
<     }
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/private-cluster/main.tf /tmp/tmp.3unHo69Lk5/workspace/modules/private-cluster/main.tf
87,88d86
<   gke_backup_agent_config = var.gke_backup_agent_config ? [{ enabled = true }] : [{ enabled = false }]
< 
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/private-cluster/variables.tf /tmp/tmp.3unHo69Lk5/workspace/modules/private-cluster/variables.tf
591,596d590
< 
< variable "gke_backup_agent_config" {
<   type        = bool
<   description = "Whether Backup for GKE agent is enabled for this cluster."
<   default     = false
< }
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/private-cluster/versions.tf /tmp/tmp.3unHo69Lk5/workspace/modules/private-cluster/versions.tf
24c24
<       version = ">= 4.47.0, < 5.0"
---
>       version = ">= 4.45.0, < 5.0"
Error: submodule's files generation has not been run, please run the
'make build' command and commit changes
 ```

**Comment by comment-bot-dev** [comment]
@Tolsee
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by comment-bot-dev** [comment]
@Tolsee
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by Tolsee** [comment]
@Jberlinsky Can you please help me understand test failure.

**Comment by ericyz** [comment]
Error Trace:	/workspace/test/integration/private_zonal_with_networking/golden.go:158
Step #75 - "verify private-zonal-with-networking":         	            				/workspace/test/integration/private_zonal_with_networking/private_zonal_with_networking_test.go:63
Step #75 - "verify private-zonal-with-networking":         	            				/workspace/test/integration/private_zonal_with_networking/terraform.go:415
Step #75 - "verify private-zonal-with-networking":         	            				/workspace/test/integration/private_zonal_with_networking/terraform.go:435
Step #75 - "verify private-zonal-with-networking":         	            				/workspace/test/integration/private_zonal_with_networking/stages.go:31
Step #75 - "verify private-zonal-with-networking":         	            				/workspace/test/integration/private_zonal_with_networking/terraform.go:435
Step #75 - "verify private-zonal-with-networking":         	Error:      	Not equal: 
Step #75 - "verify private-zonal-with-networking":         	            	expected: "{\n    \"dnsCacheConfig\": {},\n    \"gcePersistentDiskCsiDriverConfig\": {\n      \"enabled\": true\n    },\n    \"gcpFilestoreCsiDriverConfig\": {},\n    \"horizontalPodAutoscaling\": {},\n    \"httpLoadBalancing\": {},\n    \"kubernetesDashboard\": {\n      \"disabled\": true\n    },\n    \"networkPolicyConfig\": {\n      \"disabled\": true\n    }\n  }"
Step #75 - "verify private-zonal-with-networking":         	            	actual  : "{\n    \"dnsCacheConfig\": {},\n    \"gcePersistentDiskCsiDriverConfig\": {\n      \"enabled\": true\n    },\n    \"gcpFilestoreCsiDriverConfig\": {},\n    \"gkeBackupAgentConfig\": {},\n    \"horizontalPodAutoscaling\": {},\n    \"httpLoadBalancing\": {},\n    \"kubernetesDashboard\": {\n      \"disabled\": true\n    },\n    \"networkPolicyConfig\": {\n      \"disabled\": true\n    }\n  }"
Step #75 - "verify private-zonal-with-networking":         	            	
Step #75 - "verify private-zonal-with-networking":         	            	Diff:
Step #75 - "verify private-zonal-with-networking":         	            	--- Expected
Step #75 - "verify private-zonal-with-networking":         	            	+++ Actual
Step #75 - "verify private-zonal-with-networking":         	            	@@ -6,2 +6,3 @@
Step #75 - "verify private-zonal-with-networking":         	            	     "gcpFilestoreCsiDriverConfig": {},
Step #75 - "verify private-zonal-with-networking":         	            	+    "gkeBackupAgentConfig": {},
Step #75 - "verify private-zonal-with-networking":         	            	     "horizontalPodAutoscaling": {},
Step #75 - "verify private-zonal-with-networking":         	Test:       	TestPrivateZonalWithNetworking
Step #75 - "verify private-zonal-with-networking":         	Messages:   	expected addonsConfig to match fixture {
Step #75 - "verify private-zonal-with-networking":         	            	    "dnsCacheConfig": {},
Step #75 - "verify private-zonal-with-networking":         	            	    "gcePersistentDiskCsiDriverConfig": {
Step #75 - "verify private-zonal-with-networking":         	            	      "enabled": true
Step #75 - "verify private-zonal-with-networking":         	            	    },
Step #75 - "verify private-zonal-with-networking":         	            	    "gcpFilestoreCsiDriverConfig": {},
Step #75 - "verify private-zonal-with-networking":         	            	    "horizontalPodAutoscaling": {},
Step #75 - "verify private-zonal-with-networking":         	            	    "httpLoadBalancing": {},
Step #75 - "verify private-zonal-with-networking":         	            	    "kubernetesDashboard": {
Step #75 - "verify private-zonal-with-networking":         	            	      "disabled": true
Step #75 - "verify private-zonal-with-networking":         	            	    },
Step #75 - "verify private-zonal-with-networking":         	            	    "networkPolicyConfig": {
Step #75 - "verify private-zonal-with-networking":         	            	      "disabled": true
Step #75 - "verify private-zonal-with-networking":         	            	    }
Step #75 - "verify private-zonal-with-networking":         	            	  }

**Comment by ericyz** [comment]
You might want to update the https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/test/integration/private_zonal_with_networking/testdata/TestPrivateZonalWithNetworking.json to include the new attribute.

**Comment by comment-bot-dev** [comment]
@Tolsee
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by comment-bot-dev** [comment]
@Tolsee
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by comment-bot-dev** [comment]
@Tolsee
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by Tolsee** [comment]
@ericyz can you help with test failure once again? Thank you!

**Comment by apeabody** [comment]
> @ericyz can you help with test failure once again? Thank you!

Hi @Tolsee - It looks to be an ordering issue in: https://github.com/Tolsee/terraform-google-kubernetes-engine/blob/feature/promote-gke_backup_agent_config-to-ga/test/integration/private_zonal_with_networking/testdata/TestPrivateZonalWithNetworking.json#L8

Looks like you need to move `"gkeBackupAgentConfig": {}` just below `"gcpFilestoreCsiDriverConfig": {},`.

```
Step #78 - "verify private-zonal-with-networking":         	            	Diff:
Step #78 - "verify private-zonal-with-networking":         	            	--- Expected
Step #78 - "verify private-zonal-with-networking":         	            	+++ Actual
Step #78 - "verify private-zonal-with-networking":         	            	@@ -6,2 +6,3 @@
Step #78 - "verify private-zonal-with-networking":         	            	     "gcpFilestoreCsiDriverConfig": {},
Step #78 - "verify private-zonal-with-networking":         	            	+    "gkeBackupAgentConfig": {},
Step #78 - "verify private-zonal-with-networking":         	            	     "horizontalPodAutoscaling": {},
Step #78 - "verify private-zonal-with-networking":         	            	@@ -13,4 +14,3 @@
Step #78 - "verify private-zonal-with-networking":         	            	       "disabled": true
Step #78 - "verify private-zonal-with-networking":         	            	-    },
Step #78 - "verify private-zonal-with-networking":         	            	-    "gkeBackupAgentConfig": {}
Step #78 - "verify private-zonal-with-networking":         	            	+    }
```

**Comment by comment-bot-dev** [comment]
@Tolsee
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by comment-bot-dev** [comment]
@Tolsee
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by Tolsee** [comment]
@apeabody It still failing 🤦🏼. Also, I am unable to run tests locally, so can you help me once more. Thank you.

File: test/integration/private_zonal_with_networking/testdata/TestPrivateZonalWithNetworking.json [hunk 1]
```diff
@@ -50,15 +51,10 @@
   "labelFingerprint": "a9dc16a7",
   "legacyAbac": {},
   "location": "us-east4-a",
-  "locations": [
-    "us-east4-a"
-  ],
+  "locations": ["us-east4-a"],
```
Comment (apeabody) [inline test/integration/private_zonal_with_networking/testdata/TestPrivateZonalWithNetworking.json:54]: Hi @Tolsee - Looks like the original formatting was specific:
```
Step #78 - "verify private-zonal-with-networking":         	Error:      	Not equal: 
Step #78 - "verify private-zonal-with-networking":         	            	expected: "[\"us-east4-a\"]"
Step #78 - "verify private-zonal-with-networking":         	            	actual  : "[\n    \"us-east4-a\"\n  ]"
Step #78 - "verify private-zonal-with-networking":         	            	
Step #78 - "verify private-zonal-with-networking":         	            	Diff:
Step #78 - "verify private-zonal-with-networking":         	            	--- Expected
Step #78 - "verify private-zonal-with-networking":         	            	+++ Actual
Step #78 - "verify private-zonal-with-networking":         	            	@@ -1 +1,3 @@
Step #78 - "verify private-zonal-with-networking":         	            	-["us-east4-a"]
Step #78 - "verify private-zonal-with-networking":         	            	+[
Step #78 - "verify private-zonal-with-networking":         	            	+    "us-east4-a"
Step #78 - "verify private-zonal-with-networking":         	            	+  ]
Step #78 - "verify private-zonal-with-networking":         	Test:       	TestPrivateZonalWithNetworking
Step #78 - "verify private-zonal-with-networking":         	Messages:   	expected locations to match fixture ["us-east4-a"]
```

```suggestion
  "locations": [
    "us-east4-a"
  ],
```

There is a good chance this is true for the rest of the entries, so I would suggest reverting everything except the new `"gkeBackupAgentConfig": {},`.

**Review by ericyz — CHANGES_REQUESTED** [review]
Please update as Andrew' commented. LGTM for other code

**Comment by comment-bot-dev** [comment]
@Tolsee
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by comment-bot-dev** [comment]
@Tolsee
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

File: test/integration/private_zonal_with_networking/testdata/TestPrivateZonalWithNetworking.json (hunk 1 above)
Comment (Tolsee) [reply test/integration/private_zonal_with_networking/testdata/TestPrivateZonalWithNetworking.json]: Thanks updated.

**Comment by comment-bot-dev** [comment]
@Tolsee
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Review by ericyz — APPROVED** [review]
LGTM

**Pull request merged**